plasmapy/analysis/fit_functions.py

```diff
@@ -0,0 +1,1142 @@
+        -------
+        y: `numpy.ndarray`
+            Corresponding dependent variables :math:`y=f(x)` of the independent
+            variables :math:`x`.
+
+        y_err: `numpy.ndarray`
+            Uncertainties associated with the calculated dependent variables
+            :math:`\\delta y`
+        """
+        if reterr:
+            y_err, y = self.func_err(x, x_err=x_err, rety=True)
+
+            return y, y_err
+
+        y = self.func(x, *self.params)
+
+        return y
+
+    def __repr__(self):
+        return f"{self.__str__()} {self.__class__}"
+
+    @abstractmethod
+    def __str__(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def func(self, x, *args):
+        """
+        The fit function.  This signature of the function must first take the
+        independent variable followed by the parameters to be fitted as
+        separate arguments.
+
+        When sub-classing the definition should look something like::
+
+            def func(self, x, a, b, c):
+                x = self._check_x(x)
+                self._check_params(a, b, c)
+
+                return a * x ** 2 + b * x + c
+
+        Parameters
+        ----------
+        x: array_like
+            Independent variables to be passed to the fit function.
+
+        *args: Tuple[Union[float, int],...]
+            The parameters that will be adjusted to make the fit.
+
+        Returns
+        -------
+        `numpy.ndarray`:
+            The calculated dependent variables of the independent variables `x`.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def func_err(self, x, x_err=None, rety=False):
+        """
+        Calculate dependent variable uncertainties :math:`\\delta y` for
+        dependent variables :math:`y=f(x)`.
+
+        When sub-classing the definition should look something like::
+
+            def func_err(self, x, x_err=None, rety=False):
+                x = self._check_x(x)
+                if x_err is not None:
+                    x_err = self._check_x(x_err)
+
+                    if x_err.shape == ():
+                        pass
+                    elif x_err.shape != x.shape:
+                        raise ValueError(
+                            f"x_err shape {x_err.shape} must be equal the shape of "
+                            f"x {x.shape}."
+                        )
+
+                a, b, c = self.params
+                a_err, b_err, c_err = self.param_errors
+
+                # calculate error
+
+                # add x_err
+
+                if rety:
+                    y = self.func(x, a, b, c)
+                    return err, y
+
+                return err
+
+        Parameters
+        ----------
+        x: array_like
+            Independent variables to be passed to the fit function.
+
+        x_err: array_like, optional
+            Errors associated with the independent variables `x`.  Must be of
+            size one or equal to the size of `x`.
```

Comment:
```suggestion
            Errors associated with the independent variables ``x``.  Must be of
            size one or equal to the size of ``x``.
```
As I understand it, the double back ticks in ReST refer to literals whereas the single back ticks refer to code that should be interpreted.  I think we want this to be double back ticks here.